client.go

```diff
@@ -1375,3 +1397,545 @@ func (c *ToznySDKV3) UnbrokerShare(ctx context.Context, authorizerClientID strin
+
+// VerifyRawClientCredentials ensures that client credentials are in the proper form and contain all required keys
+func VerifyRawClientCredentials(credential string) error {
+	var clientConfig ClientConfigRoot
+	err := json.Unmarshal([]byte(credential), &clientConfig)
+	if err != nil {
+		return err
+	}
+	values := reflect.ValueOf(clientConfig)
+	keys := values.Type()
+	for index := 0; index < values.NumField(); index++ {
+		if keys.Field(index).Name != "ClientEmail" && values.Field(index).Interface() == "" {
+			return fmt.Errorf("Value for %s must be non-empty", keys.Field(index).Tag.Get("json"))
+		}
+	}
+	// Check that the version is either an integer or a string containing a number
+	_, versionInt := clientConfig.Version.(int)
+	_, err = strconv.Atoi(fmt.Sprintf("%v", clientConfig.Version))
+	if !versionInt && err != nil {
+		return errors.New("Value for version must be a number")
+	}
+	// ClientID must be a UUID
+	_, err = uuid.Parse(clientConfig.ClientID)
+	if err != nil {
+		return errors.New("Value for client_id should be in UUID format")
+	}
+	// Base64 encoded keys must have correct lengths
+	if len(clientConfig.PrivateSigningKey) != e3dbClients.Base64EncodedPrivateSigningKeyLength {
+		return errors.New("Invalid key length: private_signing_key")
+	}
+	if len(clientConfig.PublicSigningKey) != e3dbClients.Base64EncodedPublicSigningKeyLength {
+		return errors.New("Invalid key length: public_signing_key")
+	}
+	if len(clientConfig.PrivateKey) != e3dbClients.Base64EncodedSymmetricKeyLength {
+		return errors.New("Invalid key length: private_key")
+	}
+	if len(clientConfig.PublicKey) != e3dbClients.Base64EncodedSymmetricKeyLength {
+		return errors.New("Invalid key length: public_key")
+	}
+	return nil
+}
+
+type ListedSecrets struct {
+	List      []Secret
+	NextToken string
+}
+
+type ListSecretsOptions struct {
+	RealmName string
+	Limit     int
+	NextToken int64
+}
+
+// ListSecrets returns a list of up to limit secrets that are shared with or owned by the identity
+func (c *ToznySDKV3) ListSecrets(ctx context.Context, options ListSecretsOptions) (*ListedSecrets, error) {
+	sharedSecrets := &ListedSecrets{}
+	listRequest := storageClient.ListGroupsRequest{
+		ClientID:  uuid.MustParse(c.StorageClient.ClientID),
+		NextToken: options.NextToken,
+		Max:       options.Limit,
+	}
+	responseList, err := c.StorageClient.ListGroups(ctx, listRequest)
+	if err != nil {
+		return nil, err
+	}
+	if len(responseList.Groups) < 1 {
+		return sharedSecrets, nil
+	}
+	var found bool
+	sharedSecretList := []Secret{}
+	for index, group := range responseList.Groups {
+		if !ValidToznySecretNamespace(group.Name) {
+			continue
+		}
+		listReq := storageClient.ListGroupRecordsRequest{
+			GroupID: group.GroupID,
+		}
+		listGroupRecords, err := c.StorageClient.GetSharedWithGroup(ctx, listReq)
+		if err != nil {
+			return nil, err
+		}
+		for _, record := range listGroupRecords.ResultList {
+			found = false
+			for _, secret := range sharedSecretList {
```

Comment:
I'm not understanding how this works since `sharedSecretList` starts out as an empty slice and we only add to it....oh, okay, after reading the code for the 4th time I think its (maybe) starting to become clear to me what's going on. 

1.) naming is important, it looks like this thing you're calling `sharedSecretList` is really meant to be the container for all secrets, is that right?

2.) if blocks are for edge cases, for readability and idiomatic go code the normal path should be readable from top to bottom, that way indented paths can be ignored by the reader who is just looking to understand the happy path